ion-schema/src/main/kotlin/com/amazon/ionschema/internal/ConstraintFactoryDefault.kt

```diff
@@ -40,58 +41,55 @@ import com.amazon.ionschema.internal.constraint.TimestampPrecision
 import com.amazon.ionschema.internal.constraint.Type
 import com.amazon.ionschema.internal.constraint.Utf8ByteLength
 import com.amazon.ionschema.internal.constraint.ValidValues
-import kotlin.TODO
 
 /**
  * Default [ConstraintFactory] implementation.
- *
- * Technically, it's a constraint factory factory, but we abstract that detail away.
  */
 internal class ConstraintFactoryDefault : ConstraintFactory {
 
-    // These are the actual constraint factories.
-    private interface Constraints {
-        val newInstance: (ion: IonValue, schema: Schema) -> Constraint
-    }
-
-    private enum class Constraints_1_0(override val newInstance: (ion: IonValue, schema: Schema) -> Constraint) : Constraints {
-        all_of({ ion, schema -> AllOf(ion, schema) }),
-        annotations({ ion, schema -> Annotations(ion) }),
-        any_of({ ion, schema -> AnyOf(ion, schema) }),
-        byte_length({ ion, schema -> ByteLength(ion) }),
-        codepoint_length({ ion, schema -> CodepointLength(ion) }),
-        container_length({ ion, schema -> ContainerLength(ion) }),
-        contains({ ion, schema -> Contains(ion) }),
-        content({ ion, schema -> Content(ion) }),
-        element({ ion, schema -> Element(ion, schema) }),
-        fields({ ion, schema -> Fields(ion, schema) }),
-        not({ ion, schema -> Not(ion, schema) }),
-        occurs({ ion, schema -> OccursNoop(ion) }),
-        one_of({ ion, schema -> OneOf(ion, schema) }),
-        ordered_elements({ ion, schema -> OrderedElements(ion, schema) }),
-        precision({ ion, schema -> Precision(ion) }),
-        regex({ ion, schema -> Regex(ion) }),
-        scale({ ion, schema -> Scale(ion) }),
-        timestamp_offset({ ion, schema -> TimestampOffset(ion) }),
-        timestamp_precision({ ion, schema -> TimestampPrecision(ion) }),
-        type({ ion, schema -> Type(ion, schema) }),
-        utf8_byte_length({ ion, schema -> Utf8ByteLength(ion) }),
-        valid_values({ ion, schema -> ValidValues(ion) }),
+    /**
+     * Represents a mapping of (constraint name + supported ISL versions) to a constructor function
+     */
+    private data class ConstraintConstructor(
+        val name: String,
+        val versions: ClosedRange<IonSchemaVersion>,
+        val newInstance: (ion: IonValue, schema: Schema) -> Constraint,
+    ) {
+        constructor(name: String, versions: ClosedRange<IonSchemaVersion>, newInstance: (IonValue) -> Constraint) : this(name, versions, { ion, _ -> newInstance(ion) })
```

Comment:
(Question) Is this used to return constraint that has same functionalities for multiple versions like for `v1_0` and `v2_0` ?

Reply:
Yes. For example, `all_of` is never reasonably going to change.